### DISCUSSION TITLE
Plotter remains blank

### DISCUSSION THREAD

**Discussion by sumantkrsoni** [pyvista/pyvista#4469] [Q&A] [answered]
I am trying to generate mesh in an inverted T-shaped domain (3d) using CSV data. Unfortunately, it freezes while plotting.

Kindly suggest to plot the temperature contour using CSV data.

CSV data [x, y, z, u1, u2, u3, T]
[data.csv](https://github.com/pyvista/pyvista/files/11596636/data.csv)

Attempted code:
```python


import pyvista as pv

import pandas as pd
from pyvista.utilities import lines_from_points

dt = pd.read_csv("data.csv")
fem = dt.to_numpy()

x_fem = dt['x'].to_numpy()
y_fem = dt['y'].to_numpy()
z_fem = dt['z'].to_numpy()

T_fem = dt['T'].to_numpy()



plotter = pv.Plotter()
plotter.background_color = 'white'
for idx in range(len(x_fem)):
    cx = x_fem[idx]
    cy = y_fem[idx]
    cz = z_fem[idx]
    pts = np.column_stack((cx, cy, cz)).astype(np.float)
    lines = lines_from_points(pts)
    lines.lines = lines.faces
    tube = lines.tube(0.01).elevation()
    plotter.add_mesh(lines, show_edges=False)
plotter.show()
```

**Comment by banesullivan** [comment]
The routine you have to create lines is failing to create lines with any cells (they are single point lines... thus not a line).

Here is another approach to visualize as a point cloud:

```py
import pyvista as pv
import numpy as np
import pandas as pd
from pyvista.utilities import lines_from_points

pv.set_plot_theme('document')

dt = pd.read_csv("data.csv")
fem = dt.to_numpy()

x_fem = dt['x'].to_numpy()
y_fem = dt['y'].to_numpy()
z_fem = dt['z'].to_numpy()

T_fem = dt['T'].to_numpy()

mesh = pv.PolyData(np.c_[x_fem, y_fem, z_fem])
mesh['T'] = T_fem

plotter = pv.Plotter()
plotter.add_mesh(mesh, scalars='T')
plotter.show()
```
![screenshot](https://github.com/pyvista/pyvista/assets/22067021/483a7f1e-9c3f-405b-a5ae-340977a4c80a)


Let me know if you want to add structure to this data... I feel like lines may not be the appropriate choice, but you tell me!

**Reply by sumantkrsoni** [reply]
Thanks for your prompt response.

I have tried to plot meshing structure as shown in  
[Bunny example](https://docs.pyvista.org/version/stable/examples/02-plot/vertices.html#sphx-glr-examples-02-plot-vertices-py)
but it is appearing as displayed above.

`pl = pv.Plotter()
pl.add_mesh(
    mesh,
    show_edges=True,
    vertex_color='magenta',
    render_points_as_spheres=True,
    point_size=10,
    show_vertices=True,
)
pl.show()`

Also, kindly suggest plotting a) contours for T and , (b) streamlines for [u, v, w] from CSV file.

**Reply by banesullivan** [reply]
Where did this data come from? You'll need to export it with it's original connectivity (as opposed to just dumping out the vertices of the mesh) in order to apply contours. Otherwise, you'll have to do something like `delaunay_3d()` which just ins't going to be great

**Reply by sumantkrsoni** [reply]
Data is generated from the Freefem++ platform using following code:

`ofstream fo("data.csv");
     
      ofstream fo3("bin.csv");
      fo<<"Ra="<<Ra <<", Da ="<< Da << ", Ha ="<< Ha<<", el ="<<el<<endl;
      fo<<"x,"<<"y,"<<"z,"<<"u1,"<<"u2,"<<"u3,"<<"T"<<endl;
      for(real n=0.0; n<1.05; n=n+0.05)
      {
      for(real m=0.0; m<1.05; m=m+0.05)
      {
        for(real p=0.0; p<1.05; p=p+0.05)
        {
      if( (m>0.50 ) && ( n<0.25 || n>0.75) )
      {  fo3<<"Be strong"<<endl; }
      else{
        fo<< n << "," << m <<","<< p <<","<< u1(n,m,p)<<","<<u2(n,m,p)<<","<<u3(n,m,p)<<","<< T(n,m,p)<<endl;
      }
      }}}`

I am mainly interested in 

a) A controlled  3d mes using `elaunay_3d()` that yield a perfect 3d shape of a T-shaped domain. After that, I can approximate the results on the generated mesh.

I mean, it should not mesh the outer region of T-shaped domain.

Like attached file.

Kindly, suggest to handle it.
[mesh.pdf](https://github.com/pyvista/pyvista/files/11713088/mesh.pdf)

**Reply by banesullivan** [reply, marked as answer]
You can delaunay this point cloud like:

```py
import pyvista as pv
import numpy as np
import pandas as pd
from pyvista.utilities import lines_from_points

pv.set_plot_theme('document')

dt = pd.read_csv("data.csv")
fem = dt.to_numpy()

x_fem = dt['x'].to_numpy()
y_fem = dt['y'].to_numpy()
z_fem = dt['z'].to_numpy()

T_fem = dt['T'].to_numpy()

point_cloud = pv.PolyData(np.c_[x_fem, y_fem, z_fem])
point_cloud['T'] = T_fem

mesh = point_cloud.delaunay_3d(alpha=0.1)
mesh.plot(show_edges=True)
```

![screenshot (1)](https://github.com/pyvista/pyvista/assets/22067021/ab7707e6-58a4-4f02-bb1b-e63032285039)


But your best bet is to get the original connectivity from the starting software

**Comment by sumantkrsoni** [comment]
Thanks @banesullivan,

exactly the same plot, I was interested to plot.

Also, i feel, the appropriate value of alpha may refine this mesh further.


Thanks Again.

**Comment by sumantkrsoni** [comment]
Thanks @banesullivan for your valuable support.
